Adopt smart pointers in WebCore/xml

annevk opened the pull request:
#### 02da073676b32b2b7165aba0874e417e6cce14a3
<pre>
Adopt smart pointers in WebCore/xml
<a href="https://bugs.webkit.org/show_bug.cgi?id=294382">https://bugs.webkit.org/show_bug.cgi?id=294382</a>

Reviewed by Chris Dumez.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Canonical link: <a href="https://commits.webkit.org/296165@main">https://commits.webkit.org/296165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b87eae4048317e2b0d90181caa48ebe1c5bd1be8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107513 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112729 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58049 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109477 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27883 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81618 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22088 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96900 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62002 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21526 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15040 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57493 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91456 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15072 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115829 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25502 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90653 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34955 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93155 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90399 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23064 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35311 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13092 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30338 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34501 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40065 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34248 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37603 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1219,10 +1219,10 @@ static void restoreAttachmentElementsInFragment(DocumentFragment& fragment)
     Vector<Ref<HTMLAttachmentElement>> attachments;
-    for (auto& attachment : descendantsOfType<HTMLAttachmentElement>(fragment))
+    for (Ref attachment : descendantsOfType<HTMLAttachmentElement>(fragment))
         attachments.append(attachment);
```
`WTFMove(attachment)`

cdumez commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1240,20 +1240,20 @@ static void restoreAttachmentElementsInFragment(DocumentFragment& fragment)
 
-    for (auto& image : descendantsOfType<HTMLImageElement>(fragment))
+    for (Ref image : descendantsOfType<HTMLImageElement>(fragment))
         attachmentAssociatedElements.append(image);
```
`WTFMove(image)`

cdumez commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -1240,20 +1240,20 @@ static void restoreAttachmentElementsInFragment(DocumentFragment& fragment)
 
-    for (auto& source : descendantsOfType<HTMLSourceElement>(fragment))
+    for (Ref source : descendantsOfType<HTMLSourceElement>(fragment))
         attachmentAssociatedElements.append(source);
```
`WTFMove(source)`

cdumez commented on Source/WebCore/xml/XPathNodeSet.cpp:
```diff
@@ -201,16 +201,16 @@ void NodeSet::traversalSort() const
+    for (RefPtr node = findRootNode(m_nodes.first().get()); node; node = NodeTraversal::next(*node)) {
         if (nodes.contains(node))
-            sortedNodes.append(node);
+            sortedNodes.append(WTFMove(node));
```
This does not look OK given that `node` is used every for loop iteration: `node = NodeTraversal::next(*node)`. Now this does a use after move.

cdumez commented on Source/WebCore/xml/XPathStep.cpp:
```diff
@@ -392,27 +392,27 @@ void Step::nodesInAxis(Node& context, NodeSet& nodes) const
-            Node* node = &context;
-            if (context.isAttributeNode()) {
-                node = static_cast<Attr&>(context).ownerElement();
+            RefPtr node = &context;
```
`&` is not needed

annevk committed:
Adopt smart pointers in WebCore/xml
https://bugs.webkit.org/show_bug.cgi?id=294382

Reviewed by Chris Dumez.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Canonical link: https://commits.webkit.org/296165@main
